### ISSUE TITLE
[SCIM] dotnet 7 support

### ISSUE THREAD

**Issue by danflomin** [simpleidserver/SimpleIdServer#405]
Hello :)

I remember you wrote me about upgrading the project to dotnet 7.

Is this something you are working on?

Kind reagards
Dan

**Comment by danflomin** [comment]
When I ran my UTs with 2.0.25 and EFCore 7.0.2 I got the following exception on this lines:

```
SCIMSchema schema = await representationCommandHandler._scimSchemaCommandRepository.FindRootSCIMSchemaByResourceType(addRepresentationCommand.ResourceType);
```
<img width="863" alt="image" src="https://user-images.githubusercontent.com/33806717/217510481-529ce2cc-6438-4233-b61b-d504752e16a9.png">

The error I get is:

```
Method not found: 'Boolean Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.RemoveFromCollection(Microsoft.EntityFrameworkCore.Metadata.INavigationBase, Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry)'.
```

**Comment by simpleidserver** [comment]
Hello !

Indeed, I'm working on the release `3.0.0` which will include a lot of breaking changes in the solution 
* Merge the project UMA, OPENID, OAUTH2.0, CIBA project into one.
* Support WS-Federation.
* Migrate angular website to BLAZOR.
* Migrate SCIM project to DOTNET7

The last point `Migrate SCIM project to DOTNET7` can be done in a separate release `2.0.26`.

**Comment by simpleidserver** [comment]
The issue is fixed in the branch `master` and `release/v2.0.26`.
There was obsolete code in the `EFTransaction` class, it was used to remove orphan children in EF6 (Refer to the ticket #300).
The code is not needed anymore, because there was an issue in the EF Configuration.

When you are using different EF Version, the code is causing issue because the `RemoveCollection` function doesn't exist in `Entity` class.

You can test with the following pre-release package :

```
dotnet add package SimpleIdServer.Scim.Persistence.EFNet6 --version 2.0.26-ci-00398 --source https://www.myget.org/F/advance-ict/api/v3/index.json
```

**Comment by danflomin** [comment]
I checked the version and it looks ok but one thing.
Some attributes are created with null representation id.
To reproduce - 
1. POST user
2. POST group with the user as its member
3. PUT group with no members (remove the membership)

**Comment by simpleidserver** [comment]
The issue in EF7 is different from the one present in EF6.
* EF7 : Change tracker doesn't detect changes in the "FlatAttributes" collection, therefore the EntityState is always "Updated" and not "Deleted".
* EF6 : Change tracker detects the changes made on the "FlatAttributes" collection, but the Navigation Property "FlatAttributes" is not properly updated.

The `ForceDelete` method has been re-enabled & updated in order to support both cases.

The pre-release package is available here 

dotnet add package SimpleIdServer.Scim.Persistence.EF --version 2.0.26-ci-00398 --source https://www.myget.org/F/advance-ict/api/v3/index.json

**Comment by danflomin** [comment]
Hey 
I will check this version and update once my tests are done.

Thanks!

**Comment by danflomin** [comment]
Is this the correct version?
`2.0.26-ci-00398`

**Comment by danflomin** [comment]
Hello, 

What version should I use for testing dotnet7?

Kind regards
Dan

**Comment by danflomin** [comment]
Hello @simpleidserver ,

What version should I use for testing dotnet7?

Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello @danflomin,

You can use the Nuget Package 'SimpleIdServer.Scim.Persistence.EF --version 2.1.0-ci-00400' for testing dotnet7

KR,

SID

**Comment by danflomin** [comment]
Hello :)

I tried using version 3.0.0 and I had some issues migrating to it.

The most common issue I saw was not being able to POST a group with a member in it.

The following [line](https://github.com/simpleidserver/SimpleIdServer/blob/74bcc18a27950d10f76606e886bed05f8ea50051/src/Scim/SimpleIdServer.Scim/Commands/Handlers/AddRepresentationCommandHandler.cs#L80) fails for me.

When working with postgres I get this exception:

`25P02: current transaction is aborted, commands ignored until end of transaction block`

Did you see something like this before?


Kind regards 
Dan

**Comment by simpleidserver** [comment]
Hello,

The modifications have been tested with SQLServer but not with POSTGRES.
Unfortunately there is an issue in the library `EFCore.BulkExtensions`, Bulk insertion is not working with POSTGRES.
I'm going to take a look.

KR,

SID

**Comment by simpleidserver** [comment]
The issue is fixed after upgrading the Nuget Package `EFCore.BulkExtensions` from version 6.0.7 to 7.0.3.
Issue was linked to this one https://github.com/borisdj/EFCore.BulkExtensions/issues/931

Can-you try with this version from official NUGET feed  : 

```
dotnet add package SimpleIdServer.Scim.Persistence.EF --version 3.0.1
```

KR,

SID

**Comment by danflomin** [comment]
Thanks you very much !

I came across a different [issue](https://github.com/borisdj/EFCore.BulkExtensions/issues/1065) which holds me from upgrading to 3.0.1.

I'll wait for their response on this.


Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello,

The issue "Bulk operations with shadow property on postgresql" doesn't have an impact on the curent implementation (version 3.0.1).
In fact, bulk operations (EFCore.BulkExtensions) are executed to: 
* Insert the properties of the `SCIMRepresentationAttribute` : ValueString, SchemaAttributeId and RepresentationId.
* Remove SCIMRepresentationAttribute
* Update the properties of the `SCIMRepresentationAttribute`.

Shadow properties are not used.

While testing, Did-you encounter any kind of issue ?

KR,

SID

**Comment by danflomin** [comment]
Hello

In my usage of the library I overrided the saveChangesAsync method to add a
shadow property.

So this is why the use of the bulk library fails for me.


Kind regards
Dan

**Comment by simpleidserver** [comment]
Can-you give more details ? Explain what are you doing in the `SaveChangesAsync` ?
Are-you trying to update the property of a `SCIMRepresentationAttribute` object who is inserted via Bulk Insert ?

Because, we are using Bulk Insert to execute a lot of SQL Statements in a small amount of time, the EntityFramework Change Tracker is not aware about the changes.
Therefore, the attributes inserted via `Bulk Insert` are not tracked and cannot be updated.
Is-it the issue ? If yes, one solution can be to force the `EFCore` to track the changes like this :

```
public async Task BulkInsert(IEnumerable<SCIMRepresentationAttribute> scimRepresentationAttributes)
{
	foreach (var s in scimRepresentationAttributes)
      	_scimDbContext.Entry(s).State = EntityState.Unchanged;

      await _scimDbContext.BulkInsertAsync(scimRepresentationAttributes.ToList());
}
```

**Comment by danflomin** [comment]
I think that the issue is that the bulk library does not use the native
saveChangesAsync.

I added another column in the DB for attributes, tenat_id, as I use SCIM
for multiple tenants.
Each tenant had access only to its data.
I use the tenant ID column to separate the date using addQueryFilter in the
db Context.

During saveChangesAsync I add the tenant ID value to the shadow property.


I'm not sure if your suggestion will work, as the bulk library does not
invoke the saveChangesAsync of the db context (as far as I know).

**Comment by simpleidserver** [comment]
I checked the implementation in `EFCore.BulkExtensions`, unfortunately shadow properties are not supported by `PostgreSqlAdapter` but by `SqlServerAdapter`.
Therefore, the property `ShadowPropertyValue` cannot be used to override the shadow property :(.

A temporary solution is to configure Entity Framework to track those attributes.
Tomorrow, I will create a solution to help you.

KR,

SID

**Comment by danflomin** [comment]
That would be awesome !
Can you share with me your idea?

Just to make sure if fits my needs before you start working hard on this.


Kind regards
Dan

**Comment by danflomin** [comment]
You are awesome !

I'm going to be on vacation for the next week and a half, so it will take me some time to take a look at it and verify it is ok for me.

Thanks !

**Comment by danflomin** [comment]
Hey :)
How are you?

Sorry for the late reply.

The solution looks in the right direction, but it does not solve all my issues. 

I did the following to use it:
1. I didn't use the nuget packages, I referenced commit  `7153f70` locally from my project.
2. I changed the native EFTransaction (in `SimpleIdServer.Scim.Persistence.EF`) to be as in the custom implementation for shadow props.
3. I changed the property from  `Tenant` to `TenantId`, to work with my project.

When I: 
1. POST a user
4. POST a group with the user in it

I get the following exception message.

```
"detail": "23502: null value in column \"tenant_id\" of relation \"scim_representation_attribute_lst\" violates not-null constraint\n\nDETAIL: Failing row contains (1e790536-6b09-43d5-9fc2-9625f8e2216f, 1b41fb07-a6cd-4658-94b4-0f6fb385c492, null, df1f2280-6461-409b-8e47-ae03fd874939, 0c6695ee-3eeb-435a-b993-5b967882f969, deffa396-2b86-4814-8a1e-76d4cd6cf8ba, groups.value, 80c6856....
```

When I POST an empty group it works.


Do you know why it happens?


Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello :)

Did-you also update the method `BulkInsert` in the repository `EFSCIMRepresentationCommandRepository` with the following code ?

```
public async Task BulkInsert(IEnumerable<SCIMRepresentationAttribute> scimRepresentationAttributes)
        {
            await _scimDbContext.BulkInsertAsync(scimRepresentationAttributes.ToList());
            foreach (var scimRepresentationAttribute in scimRepresentationAttributes)
            {
                DbContext.Entry(scimRepresentationAttribute).State = EntityState.Modified;
                if(_transaction != null)
                    _transaction.BulkInsertedIds.Add(scimRepresentationAttribute.Id);
            }
        }
``` 

The full code is here : https://github.com/simpleidserver/SimpleIdServer/blob/master/samples/ScimShadowProperty/src/ScimShadowProperty/Repositories/EFSCIMRepresentationCommandRepository.cs

Don't forget to override the repositories with your custom implementation : https://github.com/simpleidserver/SimpleIdServer/blob/2726b47ae4bc1b0bd88f0c14289ec14453493a00/samples/ScimShadowProperty/src/ScimShadowProperty/Repositories/ServiceCollectionExtensions.cs

KR,

SID

**Comment by danflomin** [comment]
Hey !
Thanks for the clarification.

It looks like this solves the issue.

I took another glance at the BulkExtensions library and found out that they do support shadow properties.

Hence, we can do something like this instead (which will be a cleaner solution in my opinion).
```
        public async Task BulkInsert(IEnumerable<SCIMRepresentationAttribute> scimRepresentationAttributes)
        {
            await _scimDbContext.BulkInsertAsync(scimRepresentationAttributes.ToList(), new BulkConfig {
                EnableShadowProperties = true,
                ShadowPropertyValue = (entity, property) => {
                    if (property == "TenantId") {
                        return "tenant";
                    }

                    return property;
                }
            });
        }
```

If you think that this is a good solution, then I'd like to have an option to implement my own callback that will be used in the `BulkConfig`.

For example, have the `EFSCIMRepresentationCommandRepository` get a `IShadowProprerySetter` in the constructor, but it will use my own implementation that I will inject using IoC.

What do you think on that?


Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello,

Good suggestion :)
I create a new ticket #478 to implement this new feature.
This change will be included for the next release.

KR,

SID

**Comment by danflomin** [comment]
Hey, how have you been?

Do you have an ETA on the Notify fix and on the BulkConfig feature?
Just to know :)


Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello,

The release is planned in 3 or 4 weeks, it will include the following changes:
* Group management : DONE.
* Nuget package to extract Distributed Identity Document (DID), create Distributed Identity Document (In Progress)
* First working version of a wallet. : TODO
* Fix SCIM problems

**Comment by danflomin** [comment]
Hey, how are you?

Do we have any open issues that we should close before releasing v3.0.4 with the `BulkConfig`?


Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello,

All the issues linked to the SCIM project are fixed in the branch `release/v3.0.4` :).

The release will be published when  the specification `OPENID for Verifiable Credential Issuance` is implemented.
We made a lot of progress:
* CRUD credential template
* Support distributed identity : did:eth and did:key

Following user stores must be implemented : 
* Support pre-authorized code flow
* Support credential endpoint
* Support batch credential endpoint

**Comment by danflomin** [comment]
OK thanks :)

**Comment by danflomin** [comment]
Hello, how have you been?

I'm sorry for nagging you about this - but do you have an ETA for releasing 3.0.4?

I'm having a few issues that should be solved with this release.

I'm very excited towards this release!
So many improvements. 


Kind regards
Dan

**Comment by simpleidserver** [comment]
Hello,

Even if all the development made around "Credential Issuance" and "Distributed Identity" are almost finished.
We can make a release next friday (26 may) & we will try to include all those changes.
If we are not satisfied about the quality, we will include those changes for the next release "3.0.5".

KR,

SID

**Comment by danflomin** [comment]
Cool thanks !

**Comment by danflomin** [comment]
Thank you very much for releasing the new version !

I have a question about the new masstransit events.
When I create a user, and then POST a group with the user in it, I get 2 events for `groups.display`, one is an Added event and the other one is an Updated event.
Is the Updated event redundant?

**Comment by simpleidserver** [comment]
I create a bug ticket #518 to fix this issue.
The event must be raised only one time :)

**Comment by simpleidserver** [comment]
The release `3.0.5-rc1` is published & the bug #518 is fixed.
